examples/gate_benchmark_oplist.cpp

```diff
@@ -0,0 +1,199 @@
+#include <iostream>
+#include <map>
+#include <random>
+#include <stdexcept>
+#include <string>
+
+#include "StateVectorManaged.hpp"
+
+std::string_view strip(std::string_view str) {
+    auto start = str.find_first_not_of(" \t");
+    auto end = str.find_last_not_of(" \t");
+    return str.substr(start, end - start + 1);
+}
+
+struct GateDesc {
+    size_t n_wires;  // number of wires the gate applies to
+    size_t n_params; // number of parameters the gate requires
+};
+
+std::vector<std::pair<std::string_view, GateDesc>>
+parseGateLists(std::string_view arg) {
+    const static std::map<std::string, GateDesc> available_gates_wires = {
+        /* Single-qubit gates */
+        {"PauliX", {1, 0}},
+        {"PauliY", {1, 0}},
+        {"PauliZ", {1, 0}},
+        {"Hadamard", {1, 0}},
+        {"S", {1, 0}},
+        {"T", {1, 0}},
+        {"RX", {1, 1}},
+        {"RY", {1, 1}},
+        {"RZ", {1, 1}},
+        {"Rot", {1, 3}},
+        {"PhaseShift", {1, 1}},
+        /* Two-qubit gates */
+        {"CNOT", {2, 0}},
+        {"SWAP", {2, 0}},
+        {"ControlledPhaseShift", {2, 1}},
+        {"CRX", {2, 1}},
+        {"CRY", {2, 1}},
+        {"CRZ", {2, 1}},
+        {"CRot", {2, 3}},
+        /* Three-qubit gates */
+        {"Toffoli", {3, 0}},
+        {"CSWAP", {3, 0}}};
+
+    if (arg.empty())
+        return std::vector<std::pair<std::string_view, GateDesc>>(
+            available_gates_wires.begin(), available_gates_wires.end());
+
+    std::vector<std::pair<std::string_view, GateDesc>> ops;
+
+    if (auto pos = arg.find_first_of("["); pos != std::string_view::npos) {
+        // arg is a list "[...]"
+        auto start = pos + 1;
+        auto end = arg.find_last_of("]");
+        if (end == std::string_view::npos) {
+            throw std::invalid_argument(
+                "Argument must contain operators within square brackets [].");
+        }
+        arg = arg.substr(start, end - start);
+    }
+
+    size_t start;
+    size_t end = 0;
+    while ((start = arg.find_first_not_of(",", end)) != string::npos) {
+        end = arg.find(",", start);
+        auto op_name = strip(arg.substr(start, end - start));
+
+        auto iter = available_gates_wires.find(std::string(op_name));
+
+        if (iter == available_gates_wires.end()) {
+            std::ostringstream ss;
+            ss << "Given gate " << op_name
+               << " is not availabe"; // Change to std::format in C++20
+            throw std::invalid_argument(ss.str());
+        }
+        ops.emplace_back(*iter);
+    }
+    return ops;
+}
+
+template <typename RandomEngine>
+std::vector<size_t> generateDistinctWires(RandomEngine &re, size_t num_qubits,
+                                          size_t num_wires) {
+    std::vector<size_t> v(num_qubits, 0);
+    std::iota(v.begin(), v.end(), 0);
+    shuffle(v.begin(), v.end(), re);
+    return std::vector<size_t>(v.begin(), v.begin() + num_wires);
+}
+
+int main(int argc, char *argv[]) {
```

Comment:
Should we add a docstring of the intended feature/use of this executable?

Reply:
Added in 0522a9b47f172f2150abf9334f738cd6dc07d534